tests/test_container.py

```diff
@@ -152,4 +156,146 @@ def test_no_tag(self):
+        assert rs.content == b'some content'
+        head.assert_called_once()
+        get.assert_called_once()
+        assert i.response_cache == {'anetag': rs}
+
+    def test_all_fine_already_cached(self, get_auth, parse_www_auth, get, head):
+        rs = requests.Response()
+        rs.status_code = 200
+        rs._content = b'some content'
+        rs.headers['etag'] = 'anetag'
+        rs.headers['content-length'] = len(rs.content)
+        cache = {'anetag': rs}
+        i = Image(
+            'docker://docker.io/library/memcached:latest',
+            response_cache=cache
+        )
+        rs = requests.Response()
+        rs.status_code = 200
+        rs.headers['etag'] = 'anetag'
+        get.side_effect = Exception("Shouldn't call me")
+        head.return_value = rs
+        rs = i._request_get("myurl")
+        get.assert_not_called()
+        assert rs.content == b'some content'
+        head.assert_called_once()
+        get.assert_not_called()
+
+
+
+class TestShouldCache:
+    @staticmethod
+    def test_should_cache_small():
+        r = requests.Response()
+        r.headers['etag'] = 'abcdefg'
+        r.headers['content-length'] = '42'
+        r.status_code = 200
+        assert Image._should_cache(r)
+
+    @staticmethod
+    def test_should_cache_no_etag():
+        r = requests.Response()
+        r.headers['content-length'] = 42
+        r.status_code = 200
+        assert not Image._should_cache(r)
+
+    @staticmethod
+    def test_should_cache_humongous():
+        r = requests.Response()
+        r.headers['etag'] = 'abcd12345'
+        r.headers['content-length'] = str(10**10)
+        r.status_code = 200
+        assert not Image._should_cache(r)
+
+    @staticmethod
+    def test_should_cache_no_size():
+        r = requests.Response()
+        r.headers['etag'] = 'abcd12345'
+        r.status_code = 200
+        assert not Image._should_cache(r)
+
+    @staticmethod
+    def test_should_cache_ko():
+        r = requests.Response()
+        r.headers['etag'] = 'abcd12345'
+        r.headers['content-length'] = 42
+        r.status_code = 400
+        assert not Image._should_cache(r)
+
+
+@patch.object(requests, 'get')
+class TestRequestsGet:
+    @staticmethod
+    def request_should_not_cache(get):
+        i = Image('docker://docker.io/library/memcached:latest')
+        r = requests.Response()
+        # We don't want to test all the decorators
+        assert i.__wrapped__.__wrapped__('www.google.com') == r
+        get.assert_called_once_with('www.google.com')
+        assert not i.response_cache
+
+    @staticmethod
+    def request_should_cache(get):
+        i = Image('docker://docker.io/library/memcached:latest')
+        r = requests.Response()
+        r.headers['etag'] = 'abcd12345'
+        r.headers['content-length'] = 42
+        assert i.__wrapped__.__wrapped__('www.google.com') == r
+        get.assert_called_once_with('www.google.com')
+        assert i.response_cache['abcd12345'] == r
```

Comment:
Thank you for the tests... they are likelly to change, so I will wait the next update to review them.